hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java

```diff
@@ -0,0 +1,214 @@
+    private static final long ONE_SECOND_IN_NANOS = TimeUnit.SECONDS.toNanos(1);
+    private static final ILogger LOGGER = Logger.getLogger(SlowOperationDetector.class);
+
+    private final ConcurrentHashMap<Integer, SlowOperationLog> slowOperationLogs
+            = new ConcurrentHashMap<Integer, SlowOperationLog>();
+    private final StringBuilder stackTraceStringBuilder = new StringBuilder();
+
+    private final OperationRunner[] genericOperationRunners;
+    private final OperationRunner[] partitionOperationRunners;
+
+    private final CurrentOperationData[] genericCurrentOperationData;
+    private final CurrentOperationData[] partitionCurrentOperationData;
+
+    private final long slowOperationThresholdNanos;
+    private final long logPurgeIntervalNanos;
+    private final long logRetentionNanos;
+
+    private volatile boolean running = true;
+
+    public SlowOperationDetector(OperationRunner[] genericOperationRunners,
+                                 OperationRunner[] partitionOperationRunners,
+                                 GroupProperties groupProperties, HazelcastThreadGroup hazelcastThreadGroup) {
+        this.genericOperationRunners = genericOperationRunners;
+        this.partitionOperationRunners = partitionOperationRunners;
+
+        genericCurrentOperationData = initCurrentOperationData(genericOperationRunners);
+        partitionCurrentOperationData = initCurrentOperationData(partitionOperationRunners);
+
+        slowOperationThresholdNanos = getGroupPropertyMillis(groupProperties.SLOW_OPERATION_DETECTOR_THRESHOLD_MILLIS);
+        logPurgeIntervalNanos = getGroupPropertySeconds(groupProperties.SLOW_OPERATION_DETECTOR_LOG_PURGE_INTERVAL_SECONDS);
+        logRetentionNanos = getGroupPropertySeconds(groupProperties.SLOW_OPERATION_DETECTOR_LOG_RETENTION_SECONDS);
+
+        new Scanner(hazelcastThreadGroup).start();
+    }
+
+    public void shutdown() {
+        running = false;
+    }
+
+    Collection<SlowOperationLog> getSlowOperationLogs() {
+        return slowOperationLogs.values();
+    }
+
+    private CurrentOperationData[] initCurrentOperationData(OperationRunner[] operationRunners) {
+        CurrentOperationData[] currentOperationDataArray = new CurrentOperationData[operationRunners.length];
+        for (int i = 0; i < currentOperationDataArray.length; i++) {
+            currentOperationDataArray[i] = new CurrentOperationData();
+            currentOperationDataArray[i].operationHashCode = -1;
+        }
+        return currentOperationDataArray;
+    }
+
+    private long getGroupPropertyMillis(GroupProperties.GroupProperty groupProperty) {
+        return TimeUnit.MILLISECONDS.toNanos(groupProperty.getInteger());
+    }
+
+    private long getGroupPropertySeconds(GroupProperties.GroupProperty groupProperty) {
+        return TimeUnit.SECONDS.toNanos(groupProperty.getInteger());
+    }
+
+    private void scan(long nowNanos, long nowMillis, OperationRunner[] operationRunners,
```

Comment:
This method is too big. One small reduction; pull out the content of the for loop into a separate method.